### PR TITLE
Support swap-out in iOS and modern android app

### PR DESCRIPTION
**Pull request by dpad85**
This PR adds support for swap-out in the iOS app and in the new android app. It depends on https://github.com/ACINQ/lightning-kmp/pull/342.

### Controller

The `ScanController` controller has been  refactored a little, and the models/intents for swap-outs have been added. Also a `Parser` helper object has been created to handle data parsing.

### Android UI

image

### Choosing how to pay

A Bitcoin URI may contain a lightning payment request (in the `lightning` parameter). When that happens the wallet lets the user choose how to make the payment, on-chain or over LN.

image

### CODE REVIEW

**Comment by dpad85** [comment]
The issue with `BigDecimal` being used on iOS has been fixed.

Regarding the issue mentioned in https://github.com/ACINQ/lightning-kmp/pull/342#issuecomment-1157824884, it happens because on iOS it seems we try to send the initial amount entered by the user, not the amount after the swap-out fees. So endurance rejects the payment because it is below the amount requested by the swap invoice.

A few small issues I found when testing :

- if the payment fails, on iOS we should display the payment hash somewhere, that may be useful for debugging.
- in the details screen of the payment, the label "blockchain tx" actually points to a bitcoin address, so the wording should be fixed. That's because the server doesn't know the transaction id when the swap-out is initiated, and there's no message that would let us forward that information to Phoenix once we actually know that info. So we can only point to a Bitcoin address, which is good enough for now.

**Comment by robbiehanson** [comment]
> on iOS it seems we try to send the initial amount entered by the user, not the amount after the swap-out fees

Yup, that was it. Fixed it, and now it's working fine on iOS.

I fixed the UI issues you mentioned too. But I also thought it would be helpful if the UI displayed the amounts (swap out amount & miner fee), but I don't think we're currently storing that information anywhere ?

**Comment by robbiehanson** [comment]
I created a merge conflict with my recent refactoring commits, so I merged master in and fixed the issue.

**Comment by dpad85** [comment]
> But I also thought it would be helpful if the UI displayed the amounts (swap out amount & miner fee), but I don't think we're currently storing that information anywhere ?

Correct, the current setup does not allow us to save that information. I initially thought we could store the swap-out amount that the user actually intends to send on-chain by using the `OutgoingPayment.recipientAmount` field, but that's not possible. This field contains the amount that is received by the swap service, and includes the fee. And we can't change that without impacting the normal LN payment process, because a swap-out payment is actually a LN payment, just with some additional context. And that would be a lot of trouble.

So I think the correct solution is to amend the `OutgoingPayment.Details.SwapOut` object, and add a new `amountToChain: Satoshi` field there. Wdyt? Not sure about the naming of that field.

**Comment by dpad85** [comment]
> Maybe `minerFee: Satoshi`, and we do the math in the other direction ?

Ah good point 👍
or rather `swapFee: Satoshi` so we don't assume that the swap-out fee will always be equals to the miner fee, even if that's the case today.

**Review by robbiehanson — APPROVED** [review]

**Comment by robbiehanson** [comment]
I just realized that iOS is still missing the "choose how to pay" screen. I'll work on that next.

File: phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/data/BitcoinAddress.kt [hunk 1]
```diff
@@ -48,7 +48,12 @@ data class BitcoinAddressInfo(
         val params = Parameters.build {
             label?.let { append("label", it) }
             message?.let { append("message", it) }
-            amount?.let { append("amount", (it.toLong().toBigDecimal().movePointLeft(8).toPlainString())) } // amount field is in BTC
+            // amount field is converted to BTC
+            amount?.sat?.toString()?.padStart(9, '0')?.let {
+                val satPart = it.takeLast(8)
+                val btcPart = it.substring(0, it.length - 8)
```
[the remark this answers is not in the document]
Comment (dpad85) [reply phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/data/BitcoinAddress.kt:54]: I don't think there should be any extra 0. For example this code transform 1234 sat => 0.00001234 so we should be good!

**Comment by robbiehanson** [comment]
Screen Shot 2022-06-29 at 10 32 00

**Pull request merged**